// @condosync/web/src/pages/dashboard/DashboardPage.tsx
import { useQuery } from '@tanstack/react-query';
import { useAuthStore } from '../../store/authStore';
import { api } from '../../services/api';
import {
  Users, Package, Wrench, DollarSign, AlertTriangle,
  TrendingUp, Home, Shield, Calendar, ChevronRight,
  Loader2, Building2, Video
} from 'lucide-react';
import { formatCurrency, formatRelativeTime } from '../../lib/utils';
import { Link } from 'react-router-dom';
import {
  BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer,
  LineChart, Line, PieChart, Pie, Cell, Legend
} from 'recharts';

function StatCard({ title, value, subtitle, icon: Icon, color, trend, to }: {
  title: string;
  value: string | number;
  subtitle?: string;
  icon: React.ElementType;
  color: string;
  trend?: { value: number; label: string };
  to?: string;
}) {
  const content = (
    <div className={`bg-white rounded-xl border p-5 hover:shadow-md transition-shadow ${to ? 'cursor-pointer' : ''}`}>
      <div className="flex items-start justify-between">
        <div>
          <p className="text-sm text-muted-foreground">{title}</p>
          <p className="text-3xl font-bold mt-1">{value}</p>
          {subtitle && <p className="text-xs text-muted-foreground mt-1">{subtitle}</p>}
          {trend && (
            <div className={`flex items-center gap-1 mt-2 text-xs font-medium ${trend.value >= 0 ? 'text-green-600' : 'text-red-600'}`}>
              <TrendingUp className="w-3 h-3" />
              {trend.value >= 0 ? '+' : ''}{trend.value}% {trend.label}
            </div>
          )}
        </div>
        <div className={`${color} p-3 rounded-xl`}>
          <Icon className="w-6 h-6 text-white" />
        </div>
      </div>
    </div>
  );

  return to ? <Link to={to}>{content}</Link> : content;
}

export function DashboardPage() {
  const { selectedCondominiumId, user } = useAuthStore();

  const { data, isLoading } = useQuery({
    queryKey: ['dashboard', selectedCondominiumId],
    queryFn: async () => {
      const res = await api.get(`/dashboard/${selectedCondominiumId}`);
      return res.data.data;
    },
    enabled: !!selectedCondominiumId,
    refetchInterval: 60000, // atualizar a cada 1 min
  });

  const condominium = user?.condominiumUsers?.find(
    (cu) => cu.condominium.id === selectedCondominiumId
  )?.condominium;

  if (!selectedCondominiumId) {
    return (
      <div className="flex flex-col items-center justify-center h-64 gap-4 text-muted-foreground">
        <Building2 className="w-12 h-12" />
        <p>Selecione um condomínio para continuar</p>
      </div>
    );
  }

  if (isLoading) {
    return (
      <div className="flex items-center justify-center h-64">
        <Loader2 className="w-8 h-8 animate-spin text-blue-500" />
      </div>
    );
  }

  const d = data;

  // Séries temporais do backend
  const visitorsWeek = d?.portaria?.visitorStats || [];
  const parcelWeek = d?.portaria?.parcelStats || [];
  const financeMonths = d?.financial?.financeStats || [];

  const maintenanceStatus = [
    { name: 'Abertos', value: d?.maintenance?.openOrders || 0, color: '#3B82F6' },
    { name: 'Urgentes', value: d?.maintenance?.urgentOrders || 0, color: '#EF4444' },
  ];

  return (
    <div className="space-y-6">
      {/* Header */}
      <div>
        <h1 className="text-2xl font-bold">
          Olá, {user?.name.split(' ')[0]}! 👋
        </h1>
        <p className="text-muted-foreground">
          {condominium?.name} · {new Date().toLocaleDateString('pt-BR', { weekday: 'long', day: 'numeric', month: 'long', year: 'numeric' })}
        </p>
      </div>

      {/* Cards principais */}
      <div className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-4 gap-4">
        <StatCard
          title="Unidades Ocupadas"
          value={`${d?.summary?.occupiedUnits || 0}/${d?.summary?.totalUnits || 0}`}
          subtitle={`${d?.summary?.occupancyRate || 0}% de ocupação`}
          icon={Home}
          color="bg-blue-500"
          to="/unidades"
        />
        <StatCard
          title="Visitantes Hoje"
          value={d?.portaria?.visitorsToday || 0}
          subtitle={`${d?.portaria?.visitorsInside || 0} no momento`}
          icon={Users}
          color="bg-purple-500"
          to="/portaria/visitantes"
        />
        <StatCard
          title="Encomendas Pendentes"
          value={d?.portaria?.parcelsAwaiting || 0}
          subtitle="aguardando retirada"
          icon={Package}
          color="bg-orange-500"
          to="/portaria/encomendas"
        />
        <StatCard
          title="Chamados Abertos"
          value={d?.maintenance?.openOrders || 0}
          subtitle={`${d?.maintenance?.urgentOrders || 0} urgentes`}
          icon={Wrench}
          color="bg-red-500"
          to="/manutencao"
        />
      </div>

      {/* Segunda linha */}
      <div className="grid grid-cols-1 sm:grid-cols-3 gap-4">
        <StatCard
          title="Cobranças Pendentes"
          value={d?.financial?.pendingCharges || 0}
          subtitle={`${d?.financial?.overdueCharges || 0} em atraso`}
          icon={DollarSign}
          color="bg-green-500"
          to="/financeiro/cobranças"
        />
        <StatCard
          title="Ocorrências Abertas"
          value={d?.communication?.unreadOccurrences || 0}
          subtitle="aguardando análise"
          icon={AlertTriangle}
          color="bg-yellow-500"
          to="/comunicacao/ocorrencias"
        />
        <StatCard
          title="Reservas Confirmadas"
          value={d?.communication?.upcomingReservations || 0}
          subtitle="próximas reservas"
          icon={Calendar}
          color="bg-teal-500"
          to="/areas-comuns"
        />
      </div>

      {/* Gráficos Principais */}
      <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
        {/* Visitantes por dia */}
        <div className="bg-white rounded-xl border p-5 shadow-sm">
          <div className="flex items-center justify-between mb-4">
            <h3 className="font-semibold text-gray-700">Visitantes — Últimos 7 dias</h3>
            <Link to="/portaria/visitantes" className="text-xs text-blue-600 flex items-center gap-1 hover:underline">
              Ver detalhes <ChevronRight className="w-3 h-3" />
            </Link>
          </div>
          <ResponsiveContainer width="100%" height={200}>
            <BarChart data={visitorsWeek}>
              <CartesianGrid strokeDasharray="3 3" stroke="#f5f5f5" vertical={false} />
              <XAxis dataKey="day" axisLine={false} tickLine={false} tick={{ fontSize: 12, fill: '#6B7280' }} />
              <YAxis axisLine={false} tickLine={false} tick={{ fontSize: 12, fill: '#6B7280' }} />
              <Tooltip 
                contentStyle={{ borderRadius: '8px', border: 'none', boxShadow: '0 4px 6px -1px rgb(0 0 0 / 0.1)' }}
              />
              <Bar dataKey="visitantes" fill="#8B5CF6" radius={[4, 4, 0, 0]} />
            </BarChart>
          </ResponsiveContainer>
        </div>

        {/* Encomendas por dia */}
        <div className="bg-white rounded-xl border p-5 shadow-sm">
          <div className="flex items-center justify-between mb-4">
            <h3 className="font-semibold text-gray-700">Fluxo de Encomendas</h3>
            <Link to="/portaria/encomendas" className="text-xs text-blue-600 flex items-center gap-1 hover:underline">
              Gestão <ChevronRight className="w-3 h-3" />
            </Link>
          </div>
          <ResponsiveContainer width="100%" height={200}>
            <LineChart data={parcelWeek}>
              <CartesianGrid strokeDasharray="3 3" stroke="#f5f5f5" vertical={false} />
              <XAxis dataKey="day" axisLine={false} tickLine={false} tick={{ fontSize: 12, fill: '#6B7280' }} />
              <YAxis axisLine={false} tickLine={false} tick={{ fontSize: 12, fill: '#6B7280' }} />
              <Tooltip contentStyle={{ borderRadius: '8px', border: 'none' }} />
              <Legend verticalAlign="top" height={36} iconType="circle" />
              <Line type="monotone" dataKey="recebidas" stroke="#F59E0B" strokeWidth={2} dot={{ r: 4 }} activeDot={{ r: 6 }} />
              <Line type="monotone" dataKey="entregues" stroke="#10B981" strokeWidth={2} dot={{ r: 4 }} activeDot={{ r: 6 }} />
            </LineChart>
          </ResponsiveContainer>
        </div>
      </div>

      {/* Terceira linha: Financeiro + Comunicados */}
      <div className="grid grid-cols-1 lg:grid-cols-3 gap-6">
        {/* Saúde Financeira */}
        <div className="lg:col-span-2 bg-white rounded-xl border p-5 shadow-sm">
          <div className="flex items-center justify-between mb-4">
            <h3 className="font-semibold text-gray-700">Saúde Financeira — Últimos 6 meses</h3>
            <Link to="/financeiro" className="text-xs text-blue-600 flex items-center gap-1 hover:underline">
              Relatórios <ChevronRight className="w-3 h-3" />
            </Link>
          </div>
          <ResponsiveContainer width="100%" height={250}>
            <BarChart data={financeMonths}>
              <CartesianGrid strokeDasharray="3 3" stroke="#f5f5f5" vertical={false} />
              <XAxis dataKey="month" axisLine={false} tickLine={false} tick={{ fontSize: 12, fill: '#6B7280' }} />
              <YAxis axisLine={false} tickLine={false} tick={{ fontSize: 12, fill: '#6B7280' }} />
              <Tooltip formatter={(value: number) => formatCurrency(value)} />
              <Legend verticalAlign="top" height={36} />
              <Bar dataKey="receitas" name="Receitas" fill="#10B981" radius={[4, 4, 0, 0]} />
              <Bar dataKey="despesas" name="Despesas" fill="#EF4444" radius={[4, 4, 0, 0]} />
            </BarChart>
          </ResponsiveContainer>
        </div>

        {/* Comunicados recentes */}
        <div className="bg-white rounded-xl border p-5 shadow-sm flex flex-col">
          <div className="flex items-center justify-between mb-4">
            <h3 className="font-semibold text-gray-700">Comunicados</h3>
            <Link to="/comunicacao/avisos" className="text-xs text-blue-600 flex items-center gap-1 hover:underline">
              Ver todos <ChevronRight className="w-3 h-3" />
            </Link>
          </div>
          <div className="space-y-3 flex-1 overflow-auto">
            {d?.recentAnnouncements?.length > 0 ? (
              d.recentAnnouncements.map((a: any) => (
                <div key={a.id} className="flex gap-3 p-3 bg-gray-50 rounded-lg hover:bg-gray-100 transition-colors">
                  <div className={`w-2 h-2 rounded-full mt-1.5 shrink-0 ${a.isPinned ? 'bg-blue-500 shadow-[0_0_8px_rgba(59,130,246,0.5)]' : 'bg-gray-400'}`} />
                  <div className="min-w-0">
                    <p className="text-sm font-semibold text-gray-800 truncate">{a.title}</p>
                    <p className="text-xs text-muted-foreground mt-0.5">{formatRelativeTime(a.publishedAt)}</p>
                  </div>
                </div>
              ))
            ) : (
              <div className="flex flex-col items-center justify-center flex-1 text-muted-foreground py-8">
                <Shield className="w-8 h-8 mb-2 opacity-20" />
                <p className="text-sm">Nenhum comunicado recente</p>
              </div>
            )}
          </div>
        </div>
      </div>

      {/* Acesso rápido */}
      <div className="bg-white rounded-xl border p-5">
        <h3 className="font-semibold mb-4">Acesso Rápido</h3>
        <div className="grid grid-cols-2 sm:grid-cols-4 lg:grid-cols-6 gap-3">
          {[
            { label: 'Registrar Visita', to: '/portaria/visitantes', icon: Shield, color: 'text-purple-600 bg-purple-50' },
            { label: 'Registrar Encomenda', to: '/portaria/encomendas', icon: Package, color: 'text-orange-600 bg-orange-50' },
            { label: 'Abrir Chamado', to: '/manutencao', icon: Wrench, color: 'text-red-600 bg-red-50' },
            { label: 'Nova Cobrança', to: '/financeiro/cobranças', icon: DollarSign, color: 'text-green-600 bg-green-50' },
            { label: 'Reservar Área', to: '/areas-comuns', icon: Calendar, color: 'text-teal-600 bg-teal-50' },
            { label: 'Novo Comunicado', to: '/comunicacao/avisos', icon: Users, color: 'text-blue-600 bg-blue-50' },
            { label: 'Assembleias', to: '/assembleias', icon: Video, color: 'text-indigo-600 bg-indigo-50' },
          ].map((item) => (
            <Link
              key={item.to}
              to={item.to}
              className={`flex flex-col items-center gap-2 p-4 rounded-xl ${item.color} hover:opacity-80 transition-opacity text-center`}
            >
              <item.icon className="w-6 h-6" />
              <span className="text-xs font-medium leading-tight">{item.label}</span>
            </Link>
          ))}
        </div>
      </div>
    </div>
  );
}
